File: Java/27_Trie/H-TUF-Implement Trie II (Prefix Tree).java
// TUF: Implement Trie II (Prefix Tree)
// Trie Implementation and Advanced Operations
// https://takeuforward.org/plus/dsa/problems/trie-implementation-and-advanced-operations

/**
 * Implement "TRIE” data structure from scratch with the following functions.

• Trie(): Initialize the object of this “TRIE” data structure.
• insert(“WORD”): Insert the string “WORD” into this “TRIE” data structure.
• countWordsEqualTo(“WORD”): Return how many times this “WORD” is present in this “TRIE”.
• countWordsStartingWith(“PREFIX”): Return how many words are there in this “TRIE” that have the string “PREFIX” as a prefix.
• erase(“WORD”): Delete one occurrence of the string “WORD” from the “TRIE”.

* Examples:
Input : ["Trie", "insert", "countWordsEqualTo", "insert", "countWordsStartingWith", "erase", "countWordsStartingWith"]
[ "apple", "apple", "app", "app", "apple", "app" ]
Output : [null, null, 1, null, 2, null, 1]

Explanation :
Trie trie = new Trie()
trie.insert("apple")
trie.countWordsEqualTo("apple")  // return 1
trie.insert("app") 
trie.countWordsStartingWith("app") // return 2
trie.erase("apple")
trie.countWordsStartingWith("app")   // return 1

Input : ["Trie", "insert", "countWordsEqualTo", "insert", "erase", "countWordsStartingWith"]
[ "mango", "apple", "app", "app", "mango" ]
Output : [null, null, 0, null, null, 1]

Explanation :
Trie trie = new Trie()
trie.insert("mango")
trie.countWordsEqualTo("apple")  // return 0
trie.insert("app") 
trie.erase("app")
trie.countWordsStartingWith("mango") // return 1
 */


class Trie {
    private static class Node {
        Node[] children = new Node[26];
        int prefixCount;  // number of words passing through this node
        int endCount;     // number of words ending at this node
    }

    private final Node root;

    public Trie() {
        root = new Node();
    }

    public void insert(String word) {
        if (word == null || word.isEmpty()) return;
        Node node = root;
        for (char ch : word.toCharArray()) {
            int idx = ch - 'a';
            if (node.children[idx] == null) {
                node.children[idx] = new Node();
            }
            node = node.children[idx];
            node.prefixCount++;
        }
        node.endCount++;
    }

    public int countWordsEqualTo(String word) {
        if (word == null || word.isEmpty()) return 0;
        Node node = root;
        for (char ch : word.toCharArray()) {
            int idx = ch - 'a';
            if (node.children[idx] == null) return 0;
            node = node.children[idx];
        }
        return node.endCount;
    }

    public int countWordsStartingWith(String prefix) {
        if (prefix == null || prefix.isEmpty()) return 0;
        Node node = root;
        for (char ch : prefix.toCharArray()) {
            int idx = ch - 'a';
            if (node.children[idx] == null) return 0;
            node = node.children[idx];
        }
        return node.prefixCount;
    }

    public void erase(String word) {
        if (countWordsEqualTo(word) == 0) return; // nothing to erase
        Node node = root;
        for (char ch : word.toCharArray()) {
            int idx = ch - 'a';
            Node next = node.children[idx];
            next.prefixCount--;
            if (next.prefixCount == 0) {
                node.children[idx] = null; // prune branch to save space
                return;
            }
            node = next;
        }
        node.endCount--;
    }

    public static void main(String[] args) {
        Trie trie = new Trie();
        trie.insert("apple");
        trie.insert("app");
        System.out.println(trie.countWordsEqualTo("apple")); // 1
        System.out.println(trie.countWordsStartingWith("app")); // 2
        trie.erase("apple");
        System.out.println(trie.countWordsStartingWith("app")); // 1
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * int param_2 = obj.countWordsEqualTo(word);
 * int param_3 = obj.countWordsStartingWith(prefix);
 * obj.erase(word);
 */

File: JavaScript/27_Trie/H-TUF-Implement Trie II (Prefix Tree).js
/**
 * TUF: Implement Trie II (Prefix Tree) with additional operations
 * https://takeuforward.org/plus/dsa/problems/implement-trie-ii
 *
 * Support insert, countWordsEqualTo, countWordsStartingWith, erase
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O(L) per operation – L is word length
 * Space Complexity | O(N * L) – N words of avg length L
 */

function createTrieNode() {
    const children = {};
    let countEndsWith = 0;  // Count of words ending at this node
    let countPrefix = 0;    // Count of words with this prefix

    return {
        getChild(char) { return children[char]; },
        setChild(char, node) { children[char] = node; },
        hasChild(char) { return char in children; },
        getCountEndsWith() { return countEndsWith; },
        getCountPrefix() { return countPrefix; },
        incrementCountEndsWith() { countEndsWith++; },
        decrementCountEndsWith() { countEndsWith--; },
        incrementCountPrefix() { countPrefix++; },
        decrementCountPrefix() { countPrefix--; }
    };
}

function createTrie() {
    const root = createTrieNode();

    /**
     * Insert a word into the trie
     * @param {string} word
     */
    function insert(word) {
        let node = root;
        for (const char of word) {
            if (!node.hasChild(char)) {
                node.setChild(char, createTrieNode());
            }
            node = node.getChild(char);
            node.incrementCountPrefix();
        }
        node.incrementCountEndsWith();
    }

    /**
     * Count words equal to given word
     * @param {string} word
     * @returns {number}
     */
    function countWordsEqualTo(word) {
        let node = root;
        for (const char of word) {
            if (!node.hasChild(char)) {
                return 0;
            }
            node = node.getChild(char);
        }
        return node.getCountEndsWith();
    }

    /**
     * Count words starting with given prefix
     * @param {string} prefix
     * @returns {number}
     */
    function countWordsStartingWith(prefix) {
        let node = root;
        for (const char of prefix) {
            if (!node.hasChild(char)) {
                return 0;
            }
            node = node.getChild(char);
        }
        return node.getCountPrefix();
    }

    /**
     * Erase a word from trie (assumes word exists)
     * @param {string} word
     */
    function erase(word) {
        let node = root;
        for (const char of word) {
            if (!node.hasChild(char)) {
                return; // Word doesn't exist
            }
            node = node.getChild(char);
            node.decrementCountPrefix();
        }
        node.decrementCountEndsWith();
    }

    return { insert, countWordsEqualTo, countWordsStartingWith, erase };
}

// Test cases
const trie = createTrie();

trie.insert("apple");
trie.insert("apple");
trie.insert("apps");
trie.insert("apps");

console.log("countWordsEqualTo('apple'):", trie.countWordsEqualTo("apple")); // 2
console.log("countWordsEqualTo('apps'):", trie.countWordsEqualTo("apps")); // 2
console.log("countWordsEqualTo('app'):", trie.countWordsEqualTo("app")); // 0

console.log("countWordsStartingWith('app'):", trie.countWordsStartingWith("app")); // 4
console.log("countWordsStartingWith('appl'):", trie.countWordsStartingWith("appl")); // 2

trie.erase("apple");
console.log("\nAfter erasing one 'apple':");
console.log("countWordsEqualTo('apple'):", trie.countWordsEqualTo("apple")); // 1
console.log("countWordsStartingWith('app'):", trie.countWordsStartingWith("app")); // 3
